src/ema/math.rs

```diff
@@ -32,10 +32,9 @@ pub fn iterated_balance_ema(iterations: u32, prev: Balance, incoming: Balance, s
 pub fn iterated_volume_ema(
     iterations: u32,
     prev: (Balance, Balance, Balance, Balance),
-    incoming: (Balance, Balance, Balance, Balance),
     smoothing: Fraction,
 ) -> (Balance, Balance, Balance, Balance) {
-    volume_weighted_average(prev, incoming, exp_smoothing(smoothing, iterations))
+    volume_weighted_average(prev, (0, 0, 0, 0), exp_smoothing(smoothing, iterations))
```

Comment:
Maybe I am misunderstanding how this is being used, i.e. when the oracle is being updated.

Suppose block t includes at least one swap in the token pair in question. So then we have incoming = (v1, v2, v3, v4).

Is the volume oracle updated at the *end* of the block with (v1, v2, v3, v4)? Or is that stored until the next time the oracle is updated? I would have guessed the latter.
Suppose the latter is the case, and the next block that updates the oracle is block t+5. So this means there *was* a trade in the token pair in block t and another in block t+5, but none in t+1 ... t+4. The oracle update now must include both the (v1, v2, v3, v4) update and the blocks with zero volume. It's not hard to do and it doesn't scale with the number of empty blocks, but it's not what is currently implemented, as far as I can tell?

I'm not sure if the fact that I'm confused means we are not choosing the right boundary for what's included in HydraDX-math and what's in HydraDX-node, or if it's just going to have to be split in a weird way for oracles.

Reply:
To follow up on the last point, I think what would help me evaluate this is if there was a function something like update_ema_volume_oracle_before_swap() that just did all the math needed, in the math package.

Or, if the oracle is actually updated at the *end* of the block, it should be update_ema_volume_oracle_after_block(). That way we can fully review if we are doing the math to accurately update these oracles just by looking at the hydradx-math package.

Reply:
The oracle is always updated at the end of the block.

So, on the first ever incoming data (let's say block `b_0`), the data is used as the first oracle value.
If we get data in the block `b_0 + 1`, we do a simple weighted average at the end of the block `volume_oracle(b_0 +1) = (1-W) * volume(b_0) + W * volume(b_0 + 1)` and store it together with the the blocknumber `b_0 + 1`.
On access of the data in block `b_1`, we get the oracle value and its blocknumber. We expect the oracle to be from the parent block `b_1 - 1`. If this is not the case (`b_0 + 1 < b_1 - 1` in our example), we do a fast-forward, which will be an iterated ema with `i = b_1 - 1 - (b_0 + 1)`. This is where the volume should always approach zero over time.

An update after a period of no data starts out the same as a read: first we fast-forward to the parent block, but then we do one round of weighted average to update the value with the new incoming data.

So to summarize:

+ **Reads:** Expect value to be from the parent block; if not: fast-forward with iterated ema to the parent block.
+ **Writes:** Expect value to be from the parent block - if not also fast-foward with iterated ema - then apply the new data with a simple weighted average.

The consumers should never have to call the regular weighted average functions, only the iterated ema. (Except if they are feeling fancy and want to take the slow route by applying the regular weighted average repeatedly.)
The writers, i.e. the oracle pallet, will call (its implementation of) the iterated ema and then use the regular weighted average to compute the update on the incoming data.

Reply:
I added a test in the pallet that I think covers what you have in mind: https://github.com/galacticcouncil/warehouse/blob/0249d4bf94cf99d5f9e3402fc000e169fe369b65/ema-oracle/src/tests/invariants.rs#L119
If you want I can come up with a scenario test in the math package here as well, though.

Reply:
I see, so it sounds like the logic in hydradx-node is different from what I imagined, but it sounds to me like it's probably correct.

I guess maybe what I'd ask for then is for the functions in hydradx-math to be named more clearly to indicate what they do? The "iterated" functions do quite different things the volume versus e.g. liquidity or price oracles.

I think the function as it exists does correctly update the oracle value for a number of 0-volume blocks, but I think it should indicate precisely what it's doing.